File: python/minisgl/kernel/csrc/src/text_match.cpp
#include <minisgl/utils.h>

#include <cstdint>
#include <queue>
#include <unordered_map>
#include <vector>

#include <dlpack/dlpack.h>
#include <tvm/ffi/container/tensor.h>
#include <tvm/ffi/dtype.h>
#include <tvm/ffi/extra/c_env_api.h>
#include <tvm/ffi/function.h>
#include <tvm/ffi/object.h>

namespace {

auto is_cpu_int_tensor(const tvm::ffi::TensorView tensor, int ndim,
                       int bits) -> bool {
  return tensor.ndim() == ndim && tensor.is_contiguous() &&
         tensor.device().device_type == kDLCPU &&
         tensor.dtype().code == kDLInt && tensor.dtype().bits == bits;
}

struct Node {
  std::unordered_map<int32_t, int32_t> next;
  int32_t failure = 0;
  int32_t output_link = -1;
  std::vector<int32_t> terminal_patterns;
};

auto transition(const std::vector<Node> &trie, int32_t node,
                int32_t byte) -> int32_t {
  const auto found = trie[node].next.find(byte);
  return found == trie[node].next.end() ? -1 : found->second;
}

auto aho_find_all(const tvm::ffi::TensorView source,
                  const tvm::ffi::TensorView flat_patterns,
                  const tvm::ffi::TensorView pattern_offsets,
                  const tvm::ffi::TensorView matches,
                  int64_t max_matches) -> int64_t {
  host::RuntimeCheck(is_cpu_int_tensor(source, 1, 32),
                     "source must be a contiguous 1D CPU int32 tensor");
  host::RuntimeCheck(is_cpu_int_tensor(flat_patterns, 1, 32),
                     "flat_patterns must be a contiguous 1D CPU int32 tensor");
  host::RuntimeCheck(is_cpu_int_tensor(pattern_offsets, 1, 64),
                     "pattern_offsets must be a contiguous 1D CPU int64 tensor");
  host::RuntimeCheck(is_cpu_int_tensor(matches, 2, 64) && matches.size(1) == 3,
                     "matches must be a contiguous [N, 3] CPU int64 tensor");
  host::RuntimeCheck(max_matches > 0 && matches.size(0) >= max_matches,
                     "matches capacity is smaller than max_matches");
  host::RuntimeCheck(pattern_offsets.size(0) >= 1,
                     "pattern_offsets must contain at least zero");

  const auto *text = static_cast<const int32_t *>(source.data_ptr());
  const auto *patterns = static_cast<const int32_t *>(flat_patterns.data_ptr());
  const auto *offsets = static_cast<const int64_t *>(pattern_offsets.data_ptr());
  auto *result = static_cast<int64_t *>(matches.data_ptr());
  const int64_t pattern_count = pattern_offsets.size(0) - 1;

  std::vector<Node> trie(1);
  std::vector<int64_t> lengths(pattern_count);
  for (int64_t pattern_id = 0; pattern_id < pattern_count; ++pattern_id) {
    const int64_t begin = offsets[pattern_id];
    const int64_t end = offsets[pattern_id + 1];
    host::RuntimeCheck(begin >= 0 && end > begin &&
                           end <= flat_patterns.size(0),
                       "pattern_offsets contains an invalid or empty pattern");
    lengths[pattern_id] = end - begin;
    int32_t node = 0;
    for (int64_t pos = begin; pos < end; ++pos) {
      const int32_t byte = patterns[pos];
      host::RuntimeCheck(byte >= 0 && byte <= 255,
                         "patterns must contain byte values in [0, 255]");
      const auto found = trie[node].next.find(byte);
      if (found == trie[node].next.end()) {
        const auto child = static_cast<int32_t>(trie.size());
        trie[node].next.emplace(byte, child);
        trie.emplace_back();
        node = child;
      } else {
        node = found->second;
      }
    }
    trie[node].terminal_patterns.push_back(static_cast<int32_t>(pattern_id));
  }

  std::queue<int32_t> queue;
  for (const auto &[byte, child] : trie[0].next) {
    (void)byte;
    trie[child].failure = 0;
    queue.push(child);
  }
  while (!queue.empty()) {
    const int32_t node = queue.front();
    queue.pop();
    for (const auto &[byte, child] : trie[node].next) {
      int32_t fallback = trie[node].failure;
      int32_t fallback_child = transition(trie, fallback, byte);
      while (fallback != 0 && fallback_child == -1) {
        fallback = trie[fallback].failure;
        fallback_child = transition(trie, fallback, byte);
      }
      if (fallback_child != -1) {
        fallback = fallback_child;
      }
      trie[child].failure = fallback;
      trie[child].output_link = !trie[fallback].terminal_patterns.empty()
                                    ? fallback
                                    : trie[fallback].output_link;
      queue.push(child);
    }
  }

  int64_t count = 0;
  int32_t node = 0;
  for (int64_t pos = 0; pos < source.size(0); ++pos) {
    const int32_t byte = text[pos];
    host::RuntimeCheck(byte >= 0 && byte <= 255,
                       "source must contain byte values in [0, 255]");
    int32_t child = transition(trie, node, byte);
    while (node != 0 && child == -1) {
      node = trie[node].failure;
      child = transition(trie, node, byte);
    }
    if (child != -1) {
      node = child;
    }
    for (int32_t output_node = node; output_node != -1;
         output_node = trie[output_node].output_link) {
      for (const int32_t pattern_id : trie[output_node].terminal_patterns) {
        if (count >= max_matches) {
          return -1;
        }
        result[count * 3] = pattern_id;
        result[count * 3 + 1] = pos + 1 - lengths[pattern_id];
        result[count * 3 + 2] = pos + 1;
        ++count;
      }
    }
  }
  return count;
}

auto ordered_latest_find(const tvm::ffi::TensorView flat_sources,
                         const tvm::ffi::TensorView source_offsets,
                         const tvm::ffi::TensorView source_keys,
                         const tvm::ffi::TensorView flat_patterns,
                         const tvm::ffi::TensorView pattern_offsets,
                         const tvm::ffi::TensorView pattern_keys,
                         const tvm::ffi::TensorView matches) -> int64_t {
  host::RuntimeCheck(is_cpu_int_tensor(flat_sources, 1, 32),
                     "flat_sources must be a contiguous 1D CPU int32 tensor");
  host::RuntimeCheck(is_cpu_int_tensor(source_offsets, 1, 64),
                     "source_offsets must be a contiguous 1D CPU int64 tensor");
  host::RuntimeCheck(is_cpu_int_tensor(source_keys, 1, 64),
                     "source_keys must be a contiguous 1D CPU int64 tensor");
  host::RuntimeCheck(is_cpu_int_tensor(flat_patterns, 1, 32),
                     "flat_patterns must be a contiguous 1D CPU int32 tensor");
  host::RuntimeCheck(is_cpu_int_tensor(pattern_offsets, 1, 64),
                     "pattern_offsets must be a contiguous 1D CPU int64 tensor");
  host::RuntimeCheck(is_cpu_int_tensor(pattern_keys, 1, 64),
                     "pattern_keys must be a contiguous 1D CPU int64 tensor");
  host::RuntimeCheck(is_cpu_int_tensor(matches, 2, 64) && matches.size(1) == 3,
                     "matches must be a contiguous [N, 3] CPU int64 tensor");
  host::RuntimeCheck(source_offsets.size(0) == source_keys.size(0) + 1,
                     "source offsets and keys have inconsistent lengths");
  host::RuntimeCheck(pattern_offsets.size(0) == pattern_keys.size(0) + 1,
                     "pattern offsets and keys have inconsistent lengths");
  host::RuntimeCheck(matches.size(0) == pattern_keys.size(0),
                     "matches must provide one row per pattern");

  const auto *sources = static_cast<const int32_t *>(flat_sources.data_ptr());
  const auto *source_starts =
      static_cast<const int64_t *>(source_offsets.data_ptr());
  const auto *source_key_values =
      static_cast<const int64_t *>(source_keys.data_ptr());
  const auto *patterns = static_cast<const int32_t *>(flat_patterns.data_ptr());
  const auto *pattern_starts =
      static_cast<const int64_t *>(pattern_offsets.data_ptr());
  const auto *pattern_key_values =
      static_cast<const int64_t *>(pattern_keys.data_ptr());
  auto *result = static_cast<int64_t *>(matches.data_ptr());

  int64_t source_id = source_keys.size(0) - 1;
  for (int64_t pattern_id = pattern_keys.size(0) - 1; pattern_id >= 0;
       --pattern_id) {
    const int64_t pattern_begin = pattern_starts[pattern_id];
    const int64_t pattern_end = pattern_starts[pattern_id + 1];
    const int64_t pattern_size = pattern_end - pattern_begin;
    host::RuntimeCheck(pattern_begin >= 0 && pattern_end >= pattern_begin &&
                           pattern_end <= flat_patterns.size(0),
                       "pattern_offsets contains an invalid range");

    std::vector<int64_t> prefix(pattern_size, 0);
    for (int64_t i = 1, matched = 0; i < pattern_size; ++i) {
      const int32_t value = patterns[pattern_end - 1 - i];
      while (matched > 0 &&
             value != patterns[pattern_end - 1 - matched]) {
        matched = prefix[matched - 1];
      }
      if (value == patterns[pattern_end - 1 - matched]) {
        ++matched;
      }
      prefix[i] = matched;
    }

    bool found = false;
    while (source_id >= 0 && !found) {
      const int64_t current_source = source_id--;
      if (source_key_values[current_source] != pattern_key_values[pattern_id]) {
        continue;
      }
      const int64_t source_begin = source_starts[current_source];
      const int64_t source_end = source_starts[current_source + 1];
      host::RuntimeCheck(source_begin >= 0 && source_end >= source_begin &&
                             source_end <= flat_sources.size(0),
                         "source_offsets contains an invalid range");

      if (pattern_size == 0) {
        result[pattern_id * 3] = current_source;
        result[pattern_id * 3 + 1] = source_end - source_begin;
        result[pattern_id * 3 + 2] = source_end - source_begin;
        found = true;
        break;
      }

      int64_t matched = 0;
      for (int64_t pos = source_end; pos > source_begin; --pos) {
        const int32_t value = sources[pos - 1];
        while (matched > 0 &&
               value != patterns[pattern_end - 1 - matched]) {
          matched = prefix[matched - 1];
        }
        if (value == patterns[pattern_end - 1 - matched]) {
          ++matched;
        }
        if (matched == pattern_size) {
          const int64_t local_start = pos - source_begin - 1;
          result[pattern_id * 3] = current_source;
          result[pattern_id * 3 + 1] = local_start;
          result[pattern_id * 3 + 2] = local_start + pattern_size;
          found = true;
          break;
        }
      }
    }
    if (!found) {
      return pattern_id;
    }
  }
  return -1;
}

} // namespace

TVM_FFI_DLL_EXPORT_TYPED_FUNC(aho_find_all, aho_find_all);
TVM_FFI_DLL_EXPORT_TYPED_FUNC(ordered_latest_find, ordered_latest_find);

File: python/minisgl/kernel/text_match.py
from __future__ import annotations

import functools
from collections import deque
from typing import Sequence

import torch

from .utils import load_aot


MAX_TEXT_BYTES = 16 * 1024 * 1024
MAX_PATTERNS = 4096
MAX_PATTERN_BYTES = 1024 * 1024
MAX_MATCHES = 1_000_000


@functools.cache
def _load_text_match_module():
    return load_aot("text_match", cpp_files=["text_match.cpp"])


def _validate_inputs(text: str, patterns: Sequence[str], max_matches: int) -> None:
    if not isinstance(text, str):
        raise TypeError("text must be a string")
    if len(patterns) > MAX_PATTERNS:
        raise ValueError(f"too many text-drop patterns; maximum is {MAX_PATTERNS}")
    text_bytes = text.encode("utf-8")
    if len(text_bytes) > MAX_TEXT_BYTES:
        raise ValueError(f"text is too large for text matching; maximum is {MAX_TEXT_BYTES} bytes")
    total_pattern_bytes = 0
    for pattern in patterns:
        if not isinstance(pattern, str):
            raise TypeError("every pattern must be a string")
        if not pattern:
            raise ValueError("empty patterns are not matchable")
        total_pattern_bytes += len(pattern.encode("utf-8"))
    if total_pattern_bytes > MAX_PATTERN_BYTES:
        raise ValueError(
            "text-drop patterns are too large; "
            f"maximum combined UTF-8 size is {MAX_PATTERN_BYTES} bytes"
        )
    if max_matches <= 0 or max_matches > MAX_MATCHES:
        raise ValueError(f"max_matches must be in [1, {MAX_MATCHES}]")


def _byte_to_char_boundaries(text: str) -> dict[int, int]:
    boundaries = {0: 0}
    byte_pos = 0
    for char_pos, char in enumerate(text, start=1):
        byte_pos += len(char.encode("utf-8"))
        boundaries[byte_pos] = char_pos
    return boundaries


def _decode_byte_matches(
    text: str,
    pattern_count: int,
    matches: Sequence[tuple[int, int, int]],
) -> list[list[tuple[int, int]]]:
    boundaries = _byte_to_char_boundaries(text)
    result: list[list[tuple[int, int]]] = [[] for _ in range(pattern_count)]
    for pattern_id, byte_start, byte_end in matches:
        if byte_start not in boundaries or byte_end not in boundaries:
            raise RuntimeError("text matcher returned a span inside a UTF-8 code point")
        result[pattern_id].append((boundaries[byte_start], boundaries[byte_end]))
    for spans in result:
        spans.sort()
    return result


def find_all_reference(
    text: str,
    patterns: Sequence[str],
    *,
    max_matches: int = MAX_MATCHES,
) -> list[list[tuple[int, int]]]:
    """Aho-Corasick reference implementation with overlapping matches."""

    patterns = list(patterns)
    _validate_inputs(text, patterns, max_matches)
    if not patterns:
        return []

    encoded_patterns = [pattern.encode("utf-8") for pattern in patterns]
    transitions: list[dict[int, int]] = [{}]
    failure = [0]
    output: list[list[int]] = [[]]
    for pattern_id, pattern in enumerate(encoded_patterns):
        node = 0
        for byte in pattern:
            next_node = transitions[node].get(byte)
            if next_node is None:
                next_node = len(transitions)
                transitions[node][byte] = next_node
                transitions.append({})
                failure.append(0)
                output.append([])
            node = next_node
        output[node].append(pattern_id)

    queue: deque[int] = deque(transitions[0].values())
    while queue:
        node = queue.popleft()
        for byte, child in transitions[node].items():
            queue.append(child)
            fallback = failure[node]
            while fallback and byte not in transitions[fallback]:
                fallback = failure[fallback]
            failure[child] = transitions[fallback].get(byte, 0)
            output[child].extend(output[failure[child]])

    raw_matches: list[tuple[int, int, int]] = []
    node = 0
    for byte_end, byte in enumerate(text.encode("utf-8"), start=1):
        while node and byte not in transitions[node]:
            node = failure[node]
        node = transitions[node].get(byte, 0)
        for pattern_id in output[node]:
            byte_start = byte_end - len(encoded_patterns[pattern_id])
            raw_matches.append((pattern_id, byte_start, byte_end))
            if len(raw_matches) > max_matches:
                raise ValueError(f"text matching exceeds the maximum of {max_matches} matches")
    return _decode_byte_matches(text, len(patterns), raw_matches)


def find_all(
    text: str,
    patterns: Sequence[str],
    *,
    max_matches: int = MAX_MATCHES,
    allow_fallback: bool = True,
) -> list[list[tuple[int, int]]]:
    """Find all overlapping UTF-8 pattern occurrences in O(input + patterns + output)."""

    patterns = list(patterns)
    _validate_inputs(text, patterns, max_matches)
    if not patterns:
        return []

    source_bytes = text.encode("utf-8")
    pattern_bytes = [pattern.encode("utf-8") for pattern in patterns]
    flat_patterns = b"".join(pattern_bytes)
    offsets = [0]
    for pattern in pattern_bytes:
        offsets.append(offsets[-1] + len(pattern))

    source = torch.tensor(list(source_bytes), dtype=torch.int32, device="cpu")
    flat = torch.tensor(list(flat_patterns), dtype=torch.int32, device="cpu")
    pattern_offsets = torch.tensor(offsets, dtype=torch.int64, device="cpu")
    capacity = min(max_matches, max(1, len(source_bytes) * len(patterns)))
    output = torch.empty((capacity, 3), dtype=torch.int64, device="cpu")
    try:
        count = int(
            _load_text_match_module().aho_find_all(
                source, flat, pattern_offsets, output, int(capacity)
            )
        )
    except Exception:
        if not allow_fallback:
            raise
        return find_all_reference(text, patterns, max_matches=max_matches)
    if count < 0:
        raise ValueError(f"text matching exceeds the maximum of {max_matches} matches")
    raw = [tuple(map(int, row)) for row in output[:count].tolist()]
    return _decode_byte_matches(text, len(patterns), raw)


def _validate_ordered_inputs(
    sources: Sequence[str],
    patterns: Sequence[str],
    source_keys: Sequence[int],
    pattern_keys: Sequence[int],
) -> tuple[list[bytes], list[bytes]]:
    if len(sources) != len(source_keys):
        raise ValueError("source_keys must provide one key per source")
    if len(patterns) != len(pattern_keys):
        raise ValueError("pattern_keys must provide one key per pattern")
    if len(sources) > MAX_PATTERNS or len(patterns) > MAX_PATTERNS:
        raise ValueError(f"ordered text matching supports at most {MAX_PATTERNS} entries")
    if not all(isinstance(source, str) for source in sources):
        raise TypeError("every source must be a string")
    if not all(isinstance(pattern, str) for pattern in patterns):
        raise TypeError("every pattern must be a string")
    encoded_sources = [source.encode("utf-8") for source in sources]
    encoded_patterns = [pattern.encode("utf-8") for pattern in patterns]
    if sum(map(len, encoded_sources)) > MAX_TEXT_BYTES:
        raise ValueError(f"ordered text sources are too large; maximum is {MAX_TEXT_BYTES} bytes")
    if sum(map(len, encoded_patterns)) > MAX_PATTERN_BYTES:
        raise ValueError(
            "ordered text patterns are too large; "
            f"maximum combined UTF-8 size is {MAX_PATTERN_BYTES} bytes"
        )
    if not all(isinstance(key, int) and not isinstance(key, bool) for key in source_keys):
        raise TypeError("every source key must be an integer")
    if not all(isinstance(key, int) and not isinstance(key, bool) for key in pattern_keys):
        raise TypeError("every pattern key must be an integer")
    return encoded_sources, encoded_patterns


def _bytes_to_int_tensor(value: bytes) -> torch.Tensor:
    if not value:
        return torch.empty(0, dtype=torch.int32, device="cpu")
    return torch.frombuffer(bytearray(value), dtype=torch.uint8).to(dtype=torch.int32)


def find_ordered_latest_reference(
    sources: Sequence[str],
    patterns: Sequence[str],
    *,
    source_keys: Sequence[int],
    pattern_keys: Sequence[int],
) -> list[tuple[int, int, int]]:
    """Match ordered patterns to distinct sources, preferring the latest valid spans."""

    sources = list(sources)
    patterns = list(patterns)
    source_keys = list(source_keys)
    pattern_keys = list(pattern_keys)
    _validate_ordered_inputs(sources, patterns, source_keys, pattern_keys)

    result: list[tuple[int, int, int] | None] = [None] * len(patterns)
    source_id = len(sources) - 1
    for pattern_id in range(len(patterns) - 1, -1, -1):
        pattern = patterns[pattern_id]
        while source_id >= 0:
            if source_keys[source_id] == pattern_keys[pattern_id]:
                start = sources[source_id].rfind(pattern)
                if start >= 0:
                    result[pattern_id] = (source_id, start, start + len(pattern))
                    source_id -= 1
                    break
            source_id -= 1
        if result[pattern_id] is None:
            raise ValueError(
                f"ordered text pattern {pattern_id} has no compatible match in full_messages"
            )
    return [match for match in result if match is not None]


def find_ordered_latest(
    sources: Sequence[str],
    patterns: Sequence[str],
    *,
    source_keys: Sequence[int],
    pattern_keys: Sequence[int],
    allow_fallback: bool = True,
) -> list[tuple[int, int, int]]:
    """Linear right-to-left ordered matching with a CPU AOT KMP implementation."""

    sources = list(sources)
    patterns = list(patterns)
    source_keys = list(source_keys)
    pattern_keys = list(pattern_keys)
    source_bytes, pattern_bytes = _validate_ordered_inputs(
        sources, patterns, source_keys, pattern_keys
    )
    if not patterns:
        return []

    flat_sources = b"".join(source_bytes)
    flat_patterns = b"".join(pattern_bytes)
    source_offsets = [0]
    pattern_offsets = [0]
    for source in source_bytes:
        source_offsets.append(source_offsets[-1] + len(source))
    for pattern in pattern_bytes:
        pattern_offsets.append(pattern_offsets[-1] + len(pattern))

    source = _bytes_to_int_tensor(flat_sources)
    pattern = _bytes_to_int_tensor(flat_patterns)
    source_offset_tensor = torch.tensor(source_offsets, dtype=torch.int64, device="cpu")
    pattern_offset_tensor = torch.tensor(pattern_offsets, dtype=torch.int64, device="cpu")
    source_key_tensor = torch.tensor(source_keys, dtype=torch.int64, device="cpu")
    pattern_key_tensor = torch.tensor(pattern_keys, dtype=torch.int64, device="cpu")
    output = torch.empty((len(patterns), 3), dtype=torch.int64, device="cpu")
    try:
        missing = int(
            _load_text_match_module().ordered_latest_find(
                source,
                source_offset_tensor,
                source_key_tensor,
                pattern,
                pattern_offset_tensor,
                pattern_key_tensor,
                output,
            )
        )
    except Exception:
        if not allow_fallback:
            raise
        return find_ordered_latest_reference(
            sources,
            patterns,
            source_keys=source_keys,
            pattern_keys=pattern_keys,
        )
    if missing >= 0:
        raise ValueError(f"ordered text pattern {missing} has no compatible match in full_messages")

    byte_boundaries: dict[int, dict[int, int]] = {}
    result: list[tuple[int, int, int]] = []
    for source_id, byte_start, byte_end in output.tolist():
        source_id = int(source_id)
        byte_start = int(byte_start)
        byte_end = int(byte_end)
        boundaries = byte_boundaries.setdefault(
            source_id, _byte_to_char_boundaries(sources[source_id])
        )
        if byte_start not in boundaries or byte_end not in boundaries:
            raise RuntimeError("ordered text matcher returned a span inside a UTF-8 code point")
        result.append((source_id, boundaries[byte_start], boundaries[byte_end]))
    return result
